### PR TITLE
Hw spring 08

### PR DESCRIPTION
**Pull request by oleksandrdziubiuk**

### CODE REVIEW

File: src/main/java/com/dev/cinema/controller/InjectData.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        Role user = new Role();
+        user.setRoleName("USER");
+        roleService.add(user);
+        Role admin = new Role();
+        admin.setRoleName("ADMIN");
+        roleService.add(admin);
+
+        User userAdmin = new User();
+        userAdmin.setEmail("bob");
+        userAdmin.setPassword(passwordEncoder.encode("1234"));
```
Comment (Resci) [inline src/main/java/com/dev/cinema/controller/InjectData.java]: Why you encode password twice?

File: src/main/java/com/dev/cinema/dao/impl/RoleDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        }
+    }
+
+    @Override
+    public Optional<Role> getRoleByName(String roleName) {
+        try (Session session = sessionFactory.openSession()) {
+            Query<Role> getRole = session.createQuery("from Role "
+                    + "where role = :roleName", Role.class);
+            getRole.setParameter("roleName", roleName);
+            return Optional.ofNullable(getRole.getSingleResult());
```
Comment (Resci) [inline src/main/java/com/dev/cinema/dao/impl/RoleDaoImpl.java]: You can use `.uniqueResultOptional()`

File: src/main/java/com/dev/cinema/model/dto/UserRequestDto.java [hunk 3]
```diff
@@ -2,13 +2,17 @@
...
+import javax.validation.constraints.NotNull;
 
 @ValidPassword(field = "password", fieldMatch = "repeatPassword")
 public class UserRequestDto {
     @ValidEmail
     private String email;
     private String password;
     private String repeatPassword;
+    @NotNull
+    private List<String> roles;
```
Comment (Resci) [inline src/main/java/com/dev/cinema/model/dto/UserRequestDto.java]: You should receive roles while register user?

File: src/main/java/com/dev/cinema/dao/impl/RoleDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                session.close();
+            }
+        }
+    }
+
+    @Override
+    public Optional<Role> getRoleByName(String roleName) {
+        try (Session session = sessionFactory.openSession()) {
+            Query<Role> getRole = session.createQuery("from Role "
+                    + "where role = :roleName", Role.class);
```
Comment (Resci) [inline src/main/java/com/dev/cinema/dao/impl/RoleDaoImpl.java]: What will happen if we write "User" and we have in the database "user"?

**Review by Resci — APPROVED** [review]

File: src/main/java/com/dev/cinema/config/SecurityConfig.java [hunk 5]
```diff
@@ -1,28 +1,34 @@
...
+        this.userDetailsService = userDetailsService;
+    }
+
     @Autowired
     public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception {
         auth
-                .inMemoryAuthentication()
-                .passwordEncoder(getEncoder())
-                .withUser("user").password(getEncoder().encode("1234")).roles("USER");
+                .userDetailsService(userDetailsService);
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/config/SecurityConfig.java]: let's add passwordEncoder here too

File: src/main/java/com/dev/cinema/model/User.java [hunk 6]
```diff
@@ -17,6 +19,8 @@
     private String email;
     @Column(nullable = false)
     private String password;
+    @ManyToMany
+    private List<Role> roles;
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/model/User.java]: let's use Set instead of List here

File: src/main/java/com/dev/cinema/model/User.java [hunk 7]
```diff
@@ -42,8 +46,18 @@ public void setPassword(String password) {
         this.password = password;
     }
 
+    @ManyToMany
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/model/User.java]: should we use this annotation here? above getter?

File: src/main/java/com/dev/cinema/model/dto/UserRequestDto.java [hunk 8]
```diff
@@ -2,13 +2,15 @@
...
 import com.dev.cinema.lib.ValidPassword;
+import java.util.List;
 
 @ValidPassword(field = "password", fieldMatch = "repeatPassword")
 public class UserRequestDto {
     @ValidEmail
     private String email;
     private String password;
     private String repeatPassword;
+    private List<String> roles;
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/model/dto/UserRequestDto.java]: roles in UserRequestDto is redundant

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java [hunk 9]
```diff
@@ -1,25 +1,33 @@
...
         this.cartService = cartService;
+        this.roleService = roleService;
     }
 
     @Override
     public User register(String email, String password) {
         User user = new User();
         user.setEmail(email);
         user.setPassword(password);
+        Role role = roleService.getRoleByName("User");
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]:
```suggestion
        Role role = roleService.getRoleByName("USER");
```

File: src/main/java/com/dev/cinema/security/CustomUserDetailsService.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    private final UserService userService;
+
+    public CustomUserDetailsService(UserService userService) {
+        this.userService = userService;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        User user = userService.findByEmail(email).orElseThrow(()
+                -> new RuntimeException("Can't find user by email " + email));
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/security/CustomUserDetailsService.java]: let's throw here specific UsernameNotFoundException exception

**Review by Sofasmile — CHANGES_REQUESTED** [review]

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**